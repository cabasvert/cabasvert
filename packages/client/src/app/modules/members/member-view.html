<!--
  This file is part of CabasVert.
 
  Copyright 2017, 2018 Didier Villevalois
 
  CabasVert is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
 
  CabasVert is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
-->

<ng-container *ngIf="member.persons.length >= 1">
  <h3 *ngIf="member.persons[0] as person">
    <span text-uppercase>{{ person.lastname }}</span> {{ person.firstname }}
  </h3>
  <p *ngIf="member.persons.length > 1">
        <span *ngFor="let person of member.persons.slice(1); let isLast=last">
          <span text-uppercase>{{ person.lastname }}</span> {{ person.firstname }}{{isLast ? '' : ', '}}
        </span>
  </p>
</ng-container>
