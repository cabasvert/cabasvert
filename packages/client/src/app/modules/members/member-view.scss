p {
  font-size: 0.7rem;
  color: var(--ion-color-medium-shade);
}
