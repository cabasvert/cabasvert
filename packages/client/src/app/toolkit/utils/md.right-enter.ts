/*
 * This file is part of CabasVert.
 *
 * Copyright 2017, 2018 Didier Villevalois
 *
 * CabasVert is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CabasVert is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Animation } from '@ionic/core'

/**
 * Md Popover Enter Animation
 */
export function mdRightEnterAnimation(AnimationC: Animation, baseEl: HTMLElement, ev?: Event): Promise<Animation> {
  let originY = 'top'
  let originX = 'right'

  const contentEl = baseEl.querySelector('.popover-content') as HTMLElement
  const contentDimensions = contentEl.getBoundingClientRect()
  const contentWidth = contentDimensions.width
  const contentHeight = contentDimensions.height

  const bodyWidth = window.innerWidth
  const bodyHeight = window.innerHeight

  // If ev was passed, use that for target element
  const targetDim =
    ev && ev.target && (ev.target as HTMLElement).getBoundingClientRect()

  const targetTop =
    targetDim != null && 'top' in targetDim
      ? targetDim.top
      : bodyHeight / 2 - contentHeight / 2

  const isRTL = document.dir === 'rtl'
  const targetRight =
    targetDim != null && 'left' in targetDim
      ? isRTL
      ? targetDim.left - contentWidth + targetDim.width
      : targetDim.left
      : bodyWidth / 2 - contentWidth / 2

  const targetHeight = (targetDim && targetDim.height) || 0

  const popoverCSS: { top: any; right: any } = {
    top: targetTop,
    right: targetRight,
  }

  // If the popover right is less than the padding it is off screen
  // to the right so adjust it, else if the width of the popover
  // exceeds the body width it is off screen to the left so adjust
  if (popoverCSS.right < POPOVER_MD_BODY_PADDING) {
    popoverCSS.right = POPOVER_MD_BODY_PADDING
  } else if (
    contentWidth + POPOVER_MD_BODY_PADDING + popoverCSS.right >
    bodyWidth
  ) {
    popoverCSS.right = bodyWidth - contentWidth - POPOVER_MD_BODY_PADDING
    originX = 'left'
  }

  // If the popover when popped down stretches past bottom of screen,
  // make it pop up if there's room above
  if (
    targetTop + targetHeight + contentHeight > bodyHeight &&
    targetTop - contentHeight > 0
  ) {
    popoverCSS.top = targetTop - contentHeight
    baseEl.className = baseEl.className + ' popover-bottom'
    originY = 'bottom'
    // If there isn't room for it to pop up above the target cut it off
  } else if (targetTop + targetHeight + contentHeight > bodyHeight) {
    contentEl.style.bottom = POPOVER_MD_BODY_PADDING + 'px'
  }

  contentEl.style.top = popoverCSS.top + 'px'
  contentEl.style.right = popoverCSS.right + 'px'
  contentEl.style.transformOrigin = originY + ' ' + originX

  const baseAnimation = new AnimationC()

  const backdropAnimation = new AnimationC()
  backdropAnimation.addElement(baseEl.querySelector('ion-backdrop'))
  backdropAnimation.fromTo('opacity', 0.01, 0.32)

  const wrapperAnimation = new AnimationC()
  wrapperAnimation.addElement(baseEl.querySelector('.popover-wrapper'))
  wrapperAnimation.fromTo('opacity', 0.01, 1)

  const contentAnimation = new AnimationC()
  contentAnimation.addElement(baseEl.querySelector('.popover-content'))
  contentAnimation.fromTo('scale', 0.001, 1)

  const viewportAnimation = new AnimationC()
  viewportAnimation.addElement(baseEl.querySelector('.popover-viewport'))
  viewportAnimation.fromTo('opacity', 0.01, 1)

  return Promise.resolve(baseAnimation
    .addElement(baseEl)
    .easing('cubic-bezier(0.36,0.66,0.04,1)')
    .duration(300)
    .add(backdropAnimation)
    .add(wrapperAnimation)
    .add(contentAnimation)
    .add(viewportAnimation))
}

const POPOVER_MD_BODY_PADDING = 12
