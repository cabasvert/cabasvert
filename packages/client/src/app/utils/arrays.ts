/*
 * This file is part of CabasVert.
 *
 * Copyright 2017, 2018 Didier Villevalois
 *
 * CabasVert is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CabasVert is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
 */

declare global {
  interface Array<T> {
    contains(x: T): boolean;

    indexed(f: (t: T) => string): { [key: string]: T };

    indexedAsMap(f: (t: T) => string): Map<string, T>;

    groupBy(f: (x: T) => string): Group<T>[];
  }
}

export class Group<T> {
  constructor(public key: string, public values: T[]) {
  }
}

Array.prototype.contains = function <T>(x: T): boolean {
  return contains(this, x);
};

export function contains<T>(array: T[], elt: T) {
  return array.indexOf(elt) > -1;
}

Array.prototype.indexed = function <T>(f: (t: T) => string): { [key: string]: T } {
  return arrayIndexed(this, f);
};

export function arrayIndexed<T>(array: T[], f: (t: T) => string): { [key: string]: T } {
  return array.reduce(
    (acc, elt) => {
      acc[f(elt)] = elt;
      return acc;
    },
    {},
  );
}

Array.prototype.indexedAsMap = function <T>(f: (t: T) => string): Map<string, T> {
  return arrayIndexedAsMap(this, f);
};

export function arrayIndexedAsMap<T>(array: T[], f: (t: T) => string): Map<string, T> {
  return array.reduce(
    (acc, elt) => {
      acc.set(f(elt), elt);
      return acc;
    },
    new Map(),
  );
}

// TODO Investigate why Ionic does not see this method but WebStorm does
Array.prototype.groupBy = function <T>(f: (x: T) => string): Group<T>[] {
  return groupBy(this, f);
};

export function groupBy<T>(as: T[], f: (x: T) => string): Group<T>[] {
  let groups: { [k: string]: T[] } = {};
  as.forEach((o) => {
    let key = f(o);
    if (key in groups) groups[key].push(o);
    else groups[key] = [o];
  });
  return Object
    .keys(groups)
    .map(key => new Group(key, groups[key]))
    .reduce((acc, g) => [...acc, g], []);
}

export function copyWith<T>(source: T[], index: number, value: T): T[] {
  let copy = source.slice();
  copy[index] = value;
  return copy;
}

export function copyAdd<T>(source: T[], value: T): T[] {
  let copy = source.slice();
  copy.push(value);
  return copy;
}

export function copyRemove<T>(source: T[], index: number): T[] {
  return source.slice(0, index).concat(source.slice(index + 1, source.length));
}
