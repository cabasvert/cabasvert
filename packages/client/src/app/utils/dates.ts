/*
 * This file is part of CabasVert.
 *
 * Copyright 2017, 2018 Didier Villevalois
 *
 * CabasVert is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CabasVert is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Date {
  addDays(days: number): Date;

  substract(other: Date): number;

  getWeek(): [number, number];

  setWeek(week: [number, number]): Date;
}

Date.prototype.addDays = function (days) {
  var date = new Date(this.valueOf());
  date.setDate(date.getDate() + days);
  return date;
};

Date.prototype.substract = function (other) {
  return Math.round((this.getTime() - other.getTime()) / (1000 * 60 * 60 * 24));
};

Date.prototype.getWeek = function () {
  // Copy date so don't modify original
  var d = new Date(+this);
  d.setHours(0, 0, 0, 0);
  // Set to nearest Thursday: current date + 4 - current day number
  // Make Sunday's day number 7
  d.setDate(d.getDate() + 4 - (d.getDay() || 7));
  // Get first day of year
  var yearStart = new Date(d.getFullYear(), 0, 1);
  // Calculate full weeks to nearest Thursday
  var weekNo = Math.ceil((((d.getTime() - yearStart.getTime()) / 86400000) + 1) / 7);
  // Return array of year and week number
  return [d.getFullYear(), weekNo];
};

Date.prototype.setWeek = function (week) {
  // Copy date so don't modify original
  var d = new Date(+this);
  // Get first day of year
  var yearStart = new Date(week[0], 0, 1);
  d.setTime(yearStart.getTime() + (week[1] * 7 - 1) * 86400000);
  d.setDate(d.getDate() + this.getDay() - d.getDay());
  return d;
};
