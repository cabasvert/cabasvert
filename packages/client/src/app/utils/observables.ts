/*
 * This file is part of CabasVert.
 *
 * Copyright 2017, 2018 Didier Villevalois
 *
 * CabasVert is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CabasVert is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
 */

import { NgZone } from '@angular/core';
import {
  asyncScheduler,
  EMPTY,
  MonoTypeOperatorFunction,
  Observer,
  of,
  OperatorFunction,
  pipe,
  SchedulerLike,
  Subscription,
} from 'rxjs';
import {
  catchError,
  filter,
  ignoreElements,
  map,
  mapTo,
  observeOn,
  pairwise, subscribeOn,
  tap,
} from 'rxjs/operators';

export function previous<T>(): MonoTypeOperatorFunction<T> {
  return pipe(pairwise(), map(([p, _]) => p));
}

export function filterNotNull<T>(): MonoTypeOperatorFunction<T> {
  return filter(element => !!element);
}

export function errors<T>(): OperatorFunction<T, string> {
  return pipe(mapTo(null), ignoreElements(), catchError(e => of(e.message + '\n' + e.stack)));
}

export function ignoreErrors<T>(): MonoTypeOperatorFunction<T> {
  return catchError(_ => EMPTY);
}

export function debugObservable<T>(prefix: string, f: (T) => any = v => v): MonoTypeOperatorFunction<T> {
  return tap(new DebugObserver<T>(prefix, f));
}

class DebugObserver<T> implements Observer<T> {

  constructor(private prefix: string,
              private f: (T) => any = v => v) {
  }

  next: (value: T) => void = v => console.log(`${this.prefix} [${zoneName()}] Next value:`, this.f(v));
  error: (err: any) => void = e => console.error(`${this.prefix} [${zoneName()}] Error:`, e);
  complete: () => void = () => console.log(`${this.prefix} [${zoneName()}] Complete!`);
}

export function subscribeOutsideAngular<T>(zone: NgZone): MonoTypeOperatorFunction<T> {
  return subscribeOn(leaveZone(zone));
}

export function observeInsideAngular<T>(zone: NgZone): MonoTypeOperatorFunction<T> {
  return observeOn(enterZone(zone));
}

export function observeOutsideAngular<T>(zone: NgZone): MonoTypeOperatorFunction<T> {
  return observeOn(leaveZone(zone));
}

export function leaveZone(zone: NgZone, scheduler: SchedulerLike = asyncScheduler): SchedulerLike {
  return new LeaveZoneScheduler(zone, scheduler) as any;
}

export function enterZone(zone: NgZone, scheduler: SchedulerLike = asyncScheduler): SchedulerLike {
  return new EnterZoneScheduler(zone, scheduler) as any;
}

class LeaveZoneScheduler implements SchedulerLike {
  static now: () => number;

  constructor(private zone: NgZone, private scheduler: SchedulerLike) {
  }

  now: () => number = LeaveZoneScheduler.now;

  schedule(...args: any[]): Subscription {
    return this.zone.runOutsideAngular(() =>
      this.scheduler.schedule.apply(this.scheduler, args),
    );
  }
}

class EnterZoneScheduler implements SchedulerLike {
  static now: () => number;

  constructor(private zone: NgZone, private scheduler: SchedulerLike) {
  }

  now: () => number = EnterZoneScheduler.now;

  schedule(...args: any[]): Subscription {
    return this.zone.run(() =>
      this.scheduler.schedule.apply(this.scheduler, args),
    );
  }
}

export const zoneName: () => string = typeof Zone !== 'undefined' && Zone.current
  ? () => Zone.current.name
  : () => 'no zone';
