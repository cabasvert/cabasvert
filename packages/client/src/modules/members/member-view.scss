.item-md member-view p {
  font-size: 1.1rem;
}
