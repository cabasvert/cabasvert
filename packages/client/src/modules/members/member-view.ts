import { Component, Input } from '@angular/core'

import { Member } from "./member.model"

@Component({
  selector: 'member-view',
  templateUrl: './member-view.html',
})
export class MemberView {
  @Input() member: Member
}
