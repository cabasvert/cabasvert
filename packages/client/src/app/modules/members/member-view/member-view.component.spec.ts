/*
 * This file is part of CabasVert.
 *
 * Copyright 2017, 2018 Didier Villevalois
 *
 * CabasVert is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * CabasVert is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with CabasVert.  If not, see <http://www.gnu.org/licenses/>.
 */

import { ComponentFixture, TestBed } from '@angular/core/testing'
import { Member } from '../member.model'
import { MemberView } from './member-view'
import { setupTestBed } from '../../../utils/testbed'

describe('MemberViewComponent', () => {

  setupTestBed({
    declarations: [MemberView],
  })

  let component: MemberView
  let fixture: ComponentFixture<MemberView>

  beforeEach(() => {
    fixture = TestBed.createComponent(MemberView)
    component = fixture.componentInstance
  })

  it('should create', () => {
    expect(component).toBeTruthy()
  })

  it('should display the member with 1 person', () => {
    let member: Member = {
      _id: 'test-member',
      persons: [
        {
          firstname: 'Didier',
          lastname: 'Villevalois',
        }
      ],
    }

    component.member = member

    fixture.detectChanges()

    expect(fixture).toMatchSnapshot()
  })

  it('should display the member with 3 person', () => {
    let member: Member = {
      _id: 'test-member',
      persons: [
        {
          firstname: 'John',
          lastname: 'Smith',
        },
        {
          firstname: 'Jane',
          lastname: 'Doe',
        },
        {
          firstname: 'Lee',
          lastname: 'Smith',
        }
      ],
    }

    component.member = member

    fixture.detectChanges()

    expect(fixture).toMatchSnapshot()
  })
})
