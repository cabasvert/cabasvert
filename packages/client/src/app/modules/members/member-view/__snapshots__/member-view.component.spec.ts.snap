// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberViewComponent should display the member with 1 person 1`] = `
<member-view
  member={[Function Object]}
>
  <h3>
    <span
      text-uppercase=""
    >
      Villevalois
    </span>
     Didier 
  </h3>
</member-view>
`;

exports[`MemberViewComponent should display the member with 3 person 1`] = `
<member-view
  member={[Function Object]}
>
  <h3>
    <span
      text-uppercase=""
    >
      Smith
    </span>
     John 
  </h3><p>
    
    <span>
      <span
        text-uppercase=""
      >
        Doe
      </span>
       Jane,  
    </span>
    <span>
      <span
        text-uppercase=""
      >
        Smith
      </span>
       Lee 
    </span>
  </p>
</member-view>
`;
